File: exams/1617/translate/layout/solution/layout/horizontal-layout.cpp
#include "horizontal-layout.h"
#include <algorithm>


namespace
{
	class HorizontalLayout : public LinearLayout
	{
	public:
		HorizontalLayout(const std::vector<std::shared_ptr<Layout>>& children, double separation)
			: LinearLayout(children, separation) { }

		Box box() const override
		{
			Box result;

			result.width = std::max((int) m_children.size() - 1, 0) * m_separation;

			for (auto& child : m_children)
			{
				auto box = child->box();

				result.width += box.width;
				result.height = std::max(box.height, result.height);
			}

			return result;
		}
	};
}

std::shared_ptr<Layout> create_horizontal_layout(const std::vector<std::shared_ptr<Layout>>& children, double separation)
{
	return std::make_shared<HorizontalLayout>(children, separation);
}

File: exams/1617/translate/layout/student/java/HorizontalLayout.java
import java.util.ArrayList;


public class HorizontalLayout extends LinearLayout
{
    public HorizontalLayout(ArrayList<Layout> children, double separation)
    {
        super(children, separation);
    }

    @Override
    public Box box()
    {
        Box result = new Box();

        result.width = Math.max(children.size() - 1, 0) * separation;

        for ( Layout child : children )
        {
            Box box = child.box();

            result.width += Math.max(box.width, result.width);
            result.height = box.height;
        }

        return result;
    }

    // Te vertalen als globale functie, d.i. een functie buiten de klasse
    public static Layout createVerticalLayout(ArrayList<Layout> children, double separation)
    {
        return new VerticalLayout(children, separation);
    }
}
